src/lspclient.rs

```diff
@@ -0,0 +1,84 @@
+use std::str::FromStr;
+
+use bitcoin::secp256k1::PublicKey;
+use reqwest::Client;
+use serde::{Deserialize, Serialize};
+
+pub(crate) struct LspClient {
+    pub pubkey: PublicKey,
+    pub connection_string: String,
+    pub url: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub(crate) struct GetInfoResponse {
+    pub pubkey: String,
+    pub connection_methods: Vec<GetinfoAddress>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub(crate) struct GetinfoAddress {
+    #[serde(rename = "type")]
+    pub item_type: GetinfoAddressType,
+    #[serde(alias = "port")]
+    pub port: u16,
+    #[serde(alias = "address", skip_serializing_if = "Option::is_none")]
+    pub address: Option<String>,
+}
+
+/// Type of connection
+#[derive(Copy, Clone, Debug, Deserialize, Serialize)]
+pub(crate) enum GetinfoAddressType {
+    #[serde(rename = "dns")]
+    DNS,
+    #[serde(rename = "ipv4")]
+    IPV4,
+    #[serde(rename = "ipv6")]
+    IPV6,
+    #[serde(rename = "torv2")]
+    TORV2,
+    #[serde(rename = "torv3")]
+    TORV3,
+    #[serde(rename = "websocket")]
+    WEBSOCKET,
+}
+
+impl LspClient {
+    pub async fn new(url: String) -> anyhow::Result<Self> {
+        let http_client = Client::builder().build()?;
+        let lsp_response = http_client.get(&url).send().await?.text().await?;
+        let get_info_response: GetInfoResponse = serde_json::from_str(&lsp_response)?;
+
+        let pubkey = PublicKey::from_str(&get_info_response.pubkey)?;
+        let connection_string = get_info_response
+            .connection_methods
+            .iter()
+            .filter(|address| {
+                matches!(
+                    address.item_type,
+                    GetinfoAddressType::IPV4 | GetinfoAddressType::IPV6 | GetinfoAddressType::TORV3
+                )
+            })
+            .min_by_key(|address| match address.item_type {
+                // Prioritize IPV4, then 6, then tor
+                GetinfoAddressType::IPV4 => 0,
+                GetinfoAddressType::IPV6 => 1,
+                GetinfoAddressType::TORV3 => 2,
+                _ => unreachable!(),
+            })
+            .map(|address| {
+                format!(
+                    "{}:{}",
+                    address.address.as_ref().unwrap_or(&url),
```

Comment:
this `.unwrap_or(&url)` seems dangerous, I think it'd be better to only get the connection info from the LSP rather than sometimes replacing it